okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java

```diff
@@ -319,36 +323,73 @@ private boolean startPendingStreams() {
     this.listener = Preconditions.checkNotNull(listener, "listener");
-    // We set host to null for test.
-    if (host != null) {
-      BufferedSource source;
-      BufferedSink sink;
-      try {
-        socket = new Socket(host, port);
-        if (sslSocketFactory != null) {
-          socket = OkHttpTlsUpgrader.upgrade(
-              sslSocketFactory, socket, authorityHost, port, connectionSpec);
+
+    frameWriter = new AsyncFrameWriter(this, serializingExecutor);
+    outboundFlow = new OutboundFlowController(this, frameWriter);
+
+    // Connecting in the serializingExecutor, so that some stream operations like synStream
+    // will be executed after connected.
+    serializingExecutor.execute(new Runnable() {
+      @Override
+      public void run() {
+        if (isForTest()) {
+          clientFrameHandler = new ClientFrameHandler(testFrameReader);
+          executor.execute(clientFrameHandler);
+          connectedCallback.run();
+          frameWriter.setFrameWriter(testFrameWriter);
+          return;
+        }
+        BufferedSource source;
+        BufferedSink sink;
+        Socket sock;
+        try {
+          sock = new Socket(host, port);
+          if (sslSocketFactory != null) {
+            sock = OkHttpTlsUpgrader.upgrade(
+                sslSocketFactory, sock, authorityHost, port, connectionSpec);
+          }
+          sock.setTcpNoDelay(true);
+          source = Okio.buffer(Okio.source(sock));
+          sink = Okio.buffer(Okio.sink(sock));
+        } catch (IOException e) {
+          onIoException(e);
+          // (and probably do all of this work asynchronously instead of in calling thread)
+          throw new RuntimeException(e);
         }
-        socket.setTcpNoDelay(true);
-        source = Okio.buffer(Okio.source(socket));
-        sink = Okio.buffer(Okio.sink(socket));
-      } catch (IOException e) {
-        // TODO(jhump): should we instead notify the listener of shutdown+terminated?
-        // (and probably do all of this work asynchronously instead of in calling thread)
-        throw Status.UNAVAILABLE.withDescription("Failed connecting").withCause(e)
-            .asRuntimeException();
-      }
-      Variant variant = new Http2();
-      frameReader = variant.newReader(source, true);
-      frameWriter = new AsyncFrameWriter(variant.newWriter(sink, true), this, executor);
-      outboundFlow = new OutboundFlowController(this, frameWriter);
-      frameWriter.connectionPreface();
-      Settings settings = new Settings();
-      frameWriter.settings(settings);
-    }
 
-    clientFrameHandler = new ClientFrameHandler();
-    executor.execute(clientFrameHandler);
+        FrameWriter rawFrameWriter;
+        synchronized (lock) {
+          if (stopped) {
+            // In case user called shutdown() during the connecting.
+            try {
+              sock.close();
+            } catch (IOException ignore) {
+              // Make checkStyle happy.
```

Comment:
Why not `log.log(Level.WARNING, "Failed closing socket", e);` like in `stopIfNecessary()`?

I don't necessarily think the pre-existing all the exception handling/logging is right, but I'd rather not see the same case be handled different ways in different places.